### DISCUSSION TITLE
How to store Dio's CancelToken in database column?

### DISCUSSION THREAD

**Discussion by sharoonamjid** [simolus3/drift#1312] [Q&A] [answered]
I have a table called: `DownloadMediaData`

```
@DataClassName("DownloadMediaData")
class DownloadsMediaData extends Table {
  IntColumn get mediaId => integer()();

  TextColumn get mediaType => text()();

  TextColumn get mediaFileUrl => text()();

  TextColumn get mediaName => text()();

  // I also want to save `CancelToken `in the column

  @override
  Set<Column> get primaryKey => {mediaFileUrl};
}
```
I also want to store [dio's](https://pub.dev/packages/dio) `CancelToken` in the above table so that when I want to cancel the download request I can get that `cancelToken` from the table and cancel the specific dio request based on cancelToken.

maybe like this:

```
CancelToken cancelToken = CancelToken();

// save cancelToken in the database so that later I can cancel the download request via cancelToken if needed.
await saveCancelTokenInDownloadMediaDataInDatabase(cancelToken);

// start downloading the file and pass the created cancelToken so that we can later cancel that request through this created cancelToken
await dio.download(url, cancelToken: cancelToken);

// when needed I want to get cancelToken from the database and cancel the request of that cancelToken.
CancelToken cancelToken = await getCancelTokenOfDownloadMediaFromDatabase(3);

// cancel the download request
cancelToken.cancel('cancelled');

```
My need is how can I store `CancelToken` in the database.

Thank you.

**Comment by simolus3** [comment, marked as answer]
It's not possible to store a `CancelToken` in a database, it's not data. Data persisted in a database survives app restarts, while cancel tokens (and uploads in general) do not.

I think the cleanest approach would be to handle cancellations somewhere out of the database. You could in theory have a global map `<int, CancelToken>` somewhere. When you create a new token, you increase some counter and store it in that map. A type converter could then store the token's id an look it up as necessary. But that's really not a great solution, I wouldn't store cancel tokens in a database because they don't persist.

**Reply by sharoonamjid** [reply]
Can you please elaborate these words a bit more:

> A type converter could then store the token's id an look it up as necessary

**Reply by simolus3** [reply]
Ah sorry, actually I know realize that it probably wouldn't work. My idea was to have a  `Map<CancelToken, int>` and a `Map<int, CancelToken>` (or just a list) to look up all cancel tokens that have ever been created. You could then store the int associated with each cancellation token in the database and look it up through a type converter.

However, there's no way to implement this without storing every cancel token ever created while your app is running, so you're definitely leaking memory with that approach. Really, I would implement cancellations in a higher layer in your app.

**Reply by sharoonamjid** [reply]
I think you have directed me to the right position. Adding `CancelToken` in a list of map and later looking up that list of map would do the trick. Thank you so much. Your suggestions are always helpful 😊
